src/Components/CreateImageWizardV2/common/TypeAheadSelect.tsx

```diff
@@ -0,0 +1,390 @@
+  if (inputOptions.length === 0) {
+    return [
+      {
+        isDisabled: true,
+        children: `No results"`,
+        value: 'no results',
+      },
+    ];
+  }
+  return inputOptions.map((input) => {
+    if (typeof input === 'string') {
+      return { isDisabled: false, value: input, children: input };
+    }
+    return { isDisabled: false, value: input[0], children: input[1] };
+  });
+}
+
+function getLabel(
+  val: string | number,
+  inputOptions: [string | number, string][] | string[]
+): string {
+  if (typeof val === 'string') {
+    return val;
+  }
+  let ret: string = '';
+  inputOptions.forEach((input) => {
+    if (input[0] === val) {
+      ret = input[1];
+    }
+  });
+  return ret;
+}
+
+type TypeAheadSelectPropType<T extends string | number> = {
+  inputOptions: [T, string][] | string[];
+  fieldID: string;
+  placeholderText: string;
+  selected: [T, string] | string;
+  setSelected: Dispatch<SetStateAction<[T, string] | string>>;
+  isFetching?: boolean;
+  isError?: boolean;
+  refetch?: () => QueryActionCreatorResult<
+    QueryDefinition<
+      GetSourceListApiArg,
+      BaseQueryFn<
+        string | FetchArgs,
+        unknown,
+        FetchBaseQueryError,
+        {},
+        FetchBaseQueryMeta
+      >,
+      never,
+      V1ListSourceResponse,
+      'provisioningApi'
+    >
+  >;
+};
+
+/**
+ * Enhanced TypeAheadSelect from the pattern fly
+ * pattern https://www.patternfly.org/components/menus/select#typeahead
+ *
+ * The TypeAheadSelect component allows a user to select an element form a list
+ * of options. The user can also filter the elements by entering some text in
+ * the text input.
+ *
+ * The component can function in two different modes. Either in a mode where the
+ * key and value of the selectable items are the same, or in a mode where they
+ * and value are different. In the second operational mode, the key can be some
+ * number and the value a label of any kind.
+ *
+ * @param inputOptions a list of elements to be turned into select options.
+ * Either be a list of tuples or a list of string depending on the mode the
+ * component must operate in. If a tuple is passed, the first element of the
+ * tuple will end up being the key and the second element will be the value.
+ *
+ * @param selected the currently selected element. Either be a tuple or a string
+ * depending of the operating mode.
+ * @param setSelected a function to update the selected value.
+ *
+ * @param isFetching if provided and set to true, the component will show a
+ * loading spinner when the data is on the fly.
+ * @param isError if provided and set to true, the component will disable
+ * itself.
+ * @param refetch if provided, the component will use the refetch function to
+ * ask for a refresh of the data that makes the select options.
+ */
+function TypeAheadSelect<T extends string | number>({
+  inputOptions,
+  selected,
+  setSelected,
+  fieldID,
+  placeholderText,
+  isFetching,
+  isError,
+  refetch,
```

Comment:
I understand your comment. `TypeAheadSelct` in its current form has a couple of useful properties. The first of which is it is entirely self contained and reusable. Changing this aspect would mean that a lot of code would have to be rewritten every time we want a type ahead select. And since we're using it in at least 3 places I think this design holds a great value.

The other design you are proposing wants to have the options and their refetching and the loading spinner extracted. This design would neglect the value of having a TypeAheadSelect in my opinion. Maybe in that case it would be preferable to just use the pf5 pattern as is every time instead of doing some generalization?

Of course my preferred way holds with the designed I proposed, but we can change it ofc.

Reply:
Btw that refetch logic is in the legacy wizard. That's why I brought it there. You want to be able to refetch the sources when clicking on the text input of the typeahead. So IMO the typeahead needs to have it. It's a callback function, so i don't see it as breaching the separation of concerns since the typeahead select still don't need to have any clue about what is being refetched or if refetch will actually refetch anything.